src/libusermode/userhook.cpp

```diff
@@ -789,10 +779,13 @@ userhook::userhook(drakvuf_t drakvuf): pluginex(drakvuf, OUTPUT_DEFAULT)
 
     drakvuf_get_kernel_struct_members_array_rva(drakvuf, offset_names, __OFFSET_MAX, offsets.data());
 
+    page_mode_t pm = drakvuf_get_page_mode(drakvuf);
+    bool is32 = (pm != VMI_PM_IA32E);
+
     breakpoint_in_system_process_searcher bp;
     if (!register_trap(nullptr, protect_virtual_memory_hook_cb, bp.for_syscall_name("NtProtectVirtualMemory")) ||
         !register_trap(nullptr, map_view_of_section_hook_cb, bp.for_syscall_name("NtMapViewOfSection")) ||
-        !register_trap(nullptr, system_service_handler_hook_cb, bp.for_syscall_name("KiSystemServiceHandler")) ||
+        (!is32 && !register_trap(nullptr, system_service_handler_hook_cb, bp.for_syscall_name("KiSystemServiceHandler"))) ||
```

Comment:
This was a mechanism used to avoid VM BSOD in case we inject a wrong page fault to the VM:

![image](https://user-images.githubusercontent.com/6976513/100012475-a389d280-2dd3-11eb-8a1a-e4d3a7347a7c.png)

With your change, it's going to somehow work on x86 but this BSOD-avoiding mechanism will be not present, so in some circumstances the VM could crash very easily. Please be aware of that.

I think the actual solution would be to find out what is the equivalent of `KiSystemServiceHandler` on `x86` Windows and to implement this in a similar fashion.

Reply:
A more exact example could be that program attempts to load DLL with corrupted header such that `IMAGE_EXPORT_DIRECTORY` points to an invalid virtual address. libusermode will try to inject a page fault to force Windows to load this memory page and we will get a BSOD, due to an exception what won't be suppressed by hooking KiSystemServiceHandler

Reply:
@icedevml Thanks for the clarification. I will try.